chromium_src/base/debug/alias.h

```diff
@@ -0,0 +1,43 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_CHROMIUM_SRC_BASE_DEBUG_ALIAS_H_
+#define BRAVE_CHROMIUM_SRC_BASE_DEBUG_ALIAS_H_
+
+#include <stdint.h>
+#include <string.h>
+
+#include "base/memory/raw_ptr_exclusion.h"
+
+#include "src/base/debug/alias.h"  // IWYU pragma: export
+
+namespace base {
+namespace debug {
+
+template <typename T>
+class StackObjectCopy {
```

Comment:
After scouring everywhere for this, I finally found some concrete answers about the object model and how `StackObjectCopy` may be violating it. Unfortunately we can't have this type, and I think that makes sense why chromium doesn't offer one. The issue here is the C++ Object Model. We are invoking UB both when we `reinterpret_cast` the byte array, and when we copy the bytes to the array too[1].

> Objects of trivially-copyable types that are not potentially-overlapping subobjects are the only C++ objects that may be safely copied with std::memcpy or serialized to/from binary files with std::ofstream::write() / std::ifstream::read().

Of course, if the objects were trivially copyable, we could just copy them to the stack, and we wouldn't need this `StackObjectCopy` in the first place. However, `memcpy`ing non-trivially-copiable objects ends up subverting the language's memory model, and invoking Undefined Behaviour.

[1] https://en.cppreference.com/w/cpp/types/is_trivially_copyable

Reply:
Let's move the introduction of this helper here https://github.com/brave/brave-core/pull/26654